transmart-core-db/grails-app/services/org/transmartproject/db/userqueries/UserQuerySetService.groovy

```diff
@@ -195,14 +188,22 @@ class UserQuerySetService implements UserQuerySetResource {
                 .list() as List<QuerySetInstance>
     }
 
-    private boolean createSetWithDiffEntries(List<Long> previousPatientIds, List<Long> newPatientIds, Query query) {
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    private boolean createSetWithDiffEntries(UserQuery query) {
+        Long queryId = query.id
+        List<QuerySetInstance> previousQuerySetInstances = getInstancesForLatestQuerySet(queryId)
+        List<Long> previousPatientIds = previousQuerySetInstances.collect { it.objectId }
+
+        User user = usersResource.getUserFromUsername(query.username)
+        def queryRepresentation = UserQueryService.toRepresentation(query)
+        List<Long> newPatientIds = getPatientsForQuery(queryRepresentation, user)
 
         List<Long> addedIds = newPatientIds - previousPatientIds
         List<Long> removedIds = previousPatientIds - newPatientIds
 
         if (addedIds.size() > 0 || removedIds.size() > 0) {
             QuerySet querySet = new QuerySet()
-            querySet.query = query
+            querySet.query = Query.load(queryId)
```

Comment:
why not `query` object directly?

Reply:
`query` has `UserQuery` type here. We could choose to downcast like so `(Query) query`. But this won't work for a possible alternative implementation of `UserQuery` that is not a subclass of `Query`. e.g. Test mocks

 BTW `load` does not fetch the object from the database. It creates a proxy with id in it.

Reply:
hm, or we could simply fix the signature of the method to receive `Query` object. It's private one anyhow.